protocol/0083-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,366 @@
+
+| Status               | Benefits Enabled | Condition for entry                                       | Condition for exit                                                |
+| -------------------- | ---------------- | --------------------------------------------------------- | ----------------------------------------------------------------- |
+| `STATUS_PROPOSED`    | No               | Governance proposal valid and accepted                    | Governance proposal voting period ends (or proposal is invalid)   |
+| `STATUS_REJECTED`    | No               | Governance vote fails (or is invalid)                     | New governance proposal submitted to the network                 |
+| `STATUS_PENDING`     | No               | Governance vote passes                                    | End of epoch after network reaches proposal `enactment_timestamp` |
+| `STATUS_ACTIVE`      | Yes              | Previously `STATUS_PENDING`                               | End of epoch after network reaches proposal `closing_timestamp`   |
+| `STATUS_CLOSED`      | No               | Previously `STATUS_ACTIVE`                                | New governance proposal submitted to the network                  |
+
+## Team Mechanics
+
+A team is comprised of a referrer and all their referees. There can only ever be one referrer per team but the number of referees is unlimited.
+
+### Creating teams
+
+To create a team and generate a referral code, a party must fulfil the following criteria:
+
+- party must not currently be a **referrer**
+- party must not currently be a **referee**
+- party must be staking at least `referralProgram.minStakedVegaTokens` tokens.
+- party must not have an active liquidity provision
+
+The staking requirement is constant. If a referrer un-stakes enough tokens to fall below the requirement, they and their referees will no long be eligible for program benefits. If the referrer re-stakes enough tokens to fulfil the staking requirement, the team will become eligible for referral program benefits. Note, referees will be able to "move team" by submitting a new team.
+
+The liquidity provision restriction is constant, once a party has become a referrer they will be restricted from committing liquidity. Any liquidity provision transactions from a referrer should be rejected.
+
+To generate a team id / referral code, the party must submit a signed `CreateTeam` transaction with the following optional team information.
+
+- `enable_rewards`: a boolean value which specifies whether the team should be eligible for team based rewards (spec not started).
+- `name`: an optional team name to be added to the referral banner.
+- `teamUrl`: an optional link to a team forum, discord, etc.
+- `avatarUrl`: an optional url to an image to be used as the teams avatar
+
+```protobuf
+message CreateTeam
+    enable_rewards: True,
```

Comment:
@ValentinTrinque How much of a pain would it be if the `enable_rewards` field changed name and datatype?

UX are designing the different journeys and raised the good point that having to sign a transaction with `enable_rewards=False` isn't the clearest when this isn't to do with enabling referral rewards...

Changes would be:
- `enable_rewards` renamed to `referral_type`
- datatype changed from `bool` to an `enum` with the following names
  - `REFERRAL_TYPE_UNSPECIFIED`
  - `REFERRAL_TYPE_CODE`
  - `REFERRAL_TYPE_TEAM`

(cc: @ckpringle @mattrussell36  )

Reply:
It can be done. I will have to change the core implementation and the proto.